pkg/component/networking/nodelocaldns/nodelocaldns.go

```diff
@@ -611,3 +477,217 @@ func (n *nodeLocalDNS) getAddress(useIPv6Brackets bool) string {
+						labelKeyCleanupRequired: "true",
+					},
+					Containers: []corev1.Container{
+						{
+							Name:  "cleanup",
+							Image: n.values.AlpineImage,
+							Command: []string{
+								"/bin/sh",
+								"-c",
+								"/scripts/cleanup.sh",
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      volumeMountNameCleanUp,
+									MountPath: volumeMountPathCleanUp,
+									ReadOnly:  true,
+								},
+								{
+									Name:      volumeMountNameXtablesLock,
+									MountPath: volumeMountPathXtablesLock,
+									ReadOnly:  false,
+								},
+							},
+							ReadinessProbe: &corev1.Probe{
+								InitialDelaySeconds: 5,
+								PeriodSeconds:       2,
+								SuccessThreshold:    1,
+								FailureThreshold:    3,
+								ProbeHandler: corev1.ProbeHandler{
+									Exec: &corev1.ExecAction{
+										Command: []string{
+											"cat",
+											"/tmp/healthy",
+										},
+									},
+								},
+							},
+							SecurityContext: &corev1.SecurityContext{
+								Capabilities: &corev1.Capabilities{
+									Add: []corev1.Capability{"NET_ADMIN", "NET_RAW"},
+								},
+								Privileged: ptr.To(false),
+							},
+						},
+					},
+					Volumes: []corev1.Volume{
+						{
+							Name: volumeMountNameCleanUp,
+							VolumeSource: corev1.VolumeSource{
+								ConfigMap: &corev1.ConfigMapVolumeSource{
+									LocalObjectReference: corev1.LocalObjectReference{
+										Name: cleanupConfigMapName,
+									},
+									DefaultMode: ptr.To[int32](0775),
+								},
+							},
+						},
+						{
+							Name: volumeMountNameXtablesLock,
+							VolumeSource: corev1.VolumeSource{
+								HostPath: &corev1.HostPathVolumeSource{
+									Path: volumeMountPathXtablesLock,
+									Type: ptr.To(corev1.HostPathFileOrCreate),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	if _, err := controllerutil.CreateOrUpdate(ctx, n.values.ShootClient, cleanupDaemonSet, func() error {
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (n *nodeLocalDNS) markNodeForCleanup(ctx context.Context, shoot *gardencorev1beta1.Shoot) error {
+	// Check for node-local-dns DaemonSet existence for each worker pool
+	for _, worker := range shoot.Spec.Provider.Workers {
+		daemonSet := &appsv1.DaemonSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "node-local-dns-" + worker.Name,
+				Namespace: metav1.NamespaceSystem,
+			},
+		}
+		if err := n.values.ShootClient.Get(ctx, client.ObjectKeyFromObject(daemonSet), daemonSet); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return fmt.Errorf("failed to check DaemonSet %s: %w", daemonSet.Name, err)
+			}
+			continue
+		}
+
+		// If DaemonSet exists, add cleanup label to all nodes in the worker group
+		nodeList := &corev1.NodeList{}
+		if err := n.values.ShootClient.List(ctx, nodeList, client.MatchingLabels{
+			v1beta1constants.LabelWorkerPool: worker.Name,
+		}); err != nil {
```

Comment:
If a shoot has a lot of pools, we list again and again. The shoot client is uncached. Can you please list the nodes upfront only once and build a `map[string][]corev1.Node` with the pool names as key?